dotnet cli passing -target:Restore to MSBuild instead of -restore

As seen in: https://github.com/xamarin/net6-samples/pull/52

We have a [multi-targeted](https://github.com/xamarin/net6-samples/blob/e0259e3d94f12a9184623905d9e07c1df8567808/HelloForms/HelloForms.csproj#L3) project with:
```xml
<TargetFrameworks>net6.0-android;net6.0-ios</TargetFrameworks>
```

To deploy to an Android device/emulator (and _not_ iOS), you can run:
```cmd
> dotnet build -t:Run -p:TargetFramework=net6.0-android
```

With .NET 6.0.100-alpha.1.21064.27, this command fails with:
```
error MSB4057: The target "Run" does not exist in the project.
```

The MSBuild arguments the dotnet cli is passing seem odd:
```
-target:Restore -t:Run
```

With .NET 6.0.100-alpha.1.20562.2, it used to pass:
```
-restore -t:Run
```

In fact, you can workaround the problem by doing:
```cmd
> dotnet build -t:Run -p:TargetFramework=net6.0-android --no-restore
```

I tried creating other, simpler projects to reproduce this issue, but a console app with `TargetFrameworks=net5.0;net6.0` works fine. I always see `-restore` being passed.

I think it is something to with .NET workloads and multi-targeting at the same time?

* Example broken log: [broken.zip](https://github.com/dotnet/sdk/files/5869975/broken.zip)
* Example working log: [works.zip](https://github.com/dotnet/sdk/files/5869960/works.zip)

I still get this error with `dotnet run`:
```
> cd HelloForms
> dotnet run -f net6.0-android
Microsoft (R) Build Engine version 16.9.0-preview-21063-04+27981f15c for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

  Determining projects to restore...
  All projects are up-to-date for restore.
C:\src\net6-samples\HelloForms\HelloForms.csproj : error MSB4057: The target "Run" does not exist in the project.
```
I also tried adding `--no-restore` to the `dotnet run` command, but I'm not sure if that actually does anything.

These commands both work for me:
```
> dotnet build -t:Run -f net6.0-android --no-restore
> dotnet build -t:Run -p:TargetFramework=net6.0-android --no-restore
```

--- a/src/Cli/dotnet/commands/RestoringCommand.cs
+++ b/src/Cli/dotnet/commands/RestoringCommand.cs
@@ -56,7 +56,8 @@ namespace Microsoft.DotNet.Tools
             IEnumerable<string> restoreArguments = new string[] { "-target:Restore" };
             if (parsedArguments != null)
             {
-                restoreArguments = restoreArguments.Concat(parsedArguments.Where(a => !IsExcludedFromRestore(a)));
+                restoreArguments = restoreArguments.Concat(parsedArguments.Where(
+                    a => !IsExcludedFromRestore(a) && !IsExcludedFromSeparateRestore(a)));
             }
             if (trailingArguments != null)
             {
@@ -74,6 +75,12 @@ namespace Microsoft.DotNet.Tools
 
         private static bool IsExcludedFromRestore(string argument) 
             => argument.StartsWith("-property:TargetFramework=", StringComparison.Ordinal);
+
+        //  These arguments don't by themselves require that restore be run in a separate process,
+        //  but if there is a separate restore process they shouldn't be passed to it
+        private static bool IsExcludedFromSeparateRestore(string argument)
+            => argument.StartsWith("-t:", StringComparison.Ordinal) ||
+               argument.StartsWith("-target:", StringComparison.Ordinal);
 
         public override int Execute()
         {

--- a/src/Cli/dotnet/commands/RestoringCommand.cs
+++ b/src/Cli/dotnet/commands/RestoringCommand.cs
@@ -15,53 +15,45 @@ namespace Microsoft.DotNet.Tools
 
         public RestoringCommand(
             IEnumerable<string> msbuildArgs,
-            IEnumerable<string> parsedArguments,
-            IEnumerable<string> trailingArguments,
             bool noRestore,
             string msbuildPath = null)
-            : base(GetCommandArguments(msbuildArgs, parsedArguments, noRestore), msbuildPath)
+            : base(GetCommandArguments(msbuildArgs, noRestore), msbuildPath)
         {
-            SeparateRestoreCommand = GetSeparateRestoreCommand(parsedArguments, trailingArguments, noRestore, msbuildPath);
+            SeparateRestoreCommand = GetSeparateRestoreCommand(msbuildArgs, noRestore, msbuildPath);
         }
 
         private static IEnumerable<string> GetCommandArguments(
-            IEnumerable<string> msbuildArgs,
-            IEnumerable<string> parsedArguments,
+            IEnumerable<string> arguments,
             bool noRestore)
         {
             if (noRestore) 
             {
-                return msbuildArgs;
+                return arguments;
             }
 
-            if (HasArgumentToExcludeFromRestore(parsedArguments))
+            if (HasArgumentToExcludeFromRestore(arguments))
             {
-                return Prepend("-nologo", msbuildArgs);
+                return Prepend("-nologo", arguments);
             }
 
-            return Prepend("-restore", msbuildArgs);
+            return Prepend("-restore", arguments);
         }
 
         private static RestoreCommand GetSeparateRestoreCommand(
-            IEnumerable<string> parsedArguments,
-            IEnumerable<string> trailingArguments, 
+            IEnumerable<string> arguments,
             bool noRestore,
             string msbuildPath)
         {
-            if (noRestore || !HasArgumentToExcludeFromRestore(parsedArguments))
+            if (noRestore || !HasArgumentToExcludeFromRestore(arguments))
             {
                 return null;
             }
 
             IEnumerable<string> restoreArguments = new string[] { "-target:Restore" };
-            if (parsedArguments != null)
+            if (arguments != null)
             {
-                restoreArguments = restoreArguments.Concat(parsedArguments.Where(
+                restoreArguments = restoreArguments.Concat(arguments.Where(
                     a => !IsExcludedFromRestore(a) && !IsExcludedFromSeparateRestore(a)));
-            }
-            if (trailingArguments != null)
-            {
-                restoreArguments = restoreArguments.Concat(trailingArguments);
             }
 
             return new RestoreCommand(restoreArguments, msbuildPath);
@@ -73,14 +65,19 @@ namespace Microsoft.DotNet.Tools
         private static bool HasArgumentToExcludeFromRestore(IEnumerable<string> arguments)
             => arguments.Any(a => IsExcludedFromRestore(a));
 
-        private static bool IsExcludedFromRestore(string argument) 
-            => argument.StartsWith("-property:TargetFramework=", StringComparison.Ordinal);
+        private static readonly string[] propertyPrefixes = new string[]{ "-", "/" };
+
+        private static bool IsExcludedFromRestore(string argument)
+            => propertyPrefixes.Any(prefix => argument.StartsWith($"{prefix}property:TargetFramework=", StringComparison.Ordinal)) ||
+               propertyPrefixes.Any(prefix => argument.StartsWith($"{prefix}p:TargetFramework=", StringComparison.Ordinal));
 
         //  These arguments don't by themselves require that restore be run in a separate process,
         //  but if there is a separate restore process they shouldn't be passed to it
         private static bool IsExcludedFromSeparateRestore(string argument)
-            => argument.StartsWith("-t:", StringComparison.Ordinal) ||
-               argument.StartsWith("-target:", StringComparison.Ordinal);
+            => propertyPrefixes.Any(prefix => argument.StartsWith($"{prefix}t:", StringComparison.Ordinal)) ||
+               propertyPrefixes.Any(prefix => argument.StartsWith($"{prefix}target:", StringComparison.Ordinal)) ||
+               propertyPrefixes.Any(prefix => argument.StartsWith($"{prefix}consoleloggerparameters:", StringComparison.Ordinal)) ||
+               propertyPrefixes.Any(prefix => argument.StartsWith($"{prefix}clp:", StringComparison.Ordinal));
 
         public override int Execute()
         {

--- a/src/Cli/dotnet/commands/dotnet-build/BuildCommand.cs
+++ b/src/Cli/dotnet/commands/dotnet-build/BuildCommand.cs
@@ -15,11 +15,9 @@ namespace Microsoft.DotNet.Tools.Build
     {
         public BuildCommand(
             IEnumerable<string> msbuildArgs,
-            IEnumerable<string> userDefinedArguments,
-            IEnumerable<string> trailingArguments,
             bool noRestore,
             string msbuildPath = null)
-            : base(msbuildArgs, userDefinedArguments, trailingArguments, noRestore, msbuildPath)
+            : base(msbuildArgs, noRestore, msbuildPath)
         {
         }
 
@@ -51,8 +49,6 @@ namespace Microsoft.DotNet.Tools.Build
 
             BuildCommand command = new BuildCommand(
                 msbuildArgs,
-                parseResult.OptionValuesToBeForwarded(BuildCommandParser.GetCommand()),
-                arguments,
                 noRestore,
                 msbuildPath);
 

--- a/src/Cli/dotnet/commands/dotnet-pack/PackCommand.cs
+++ b/src/Cli/dotnet/commands/dotnet-pack/PackCommand.cs
@@ -4,10 +4,7 @@
 using System.Collections.Generic;
 using Microsoft.DotNet.Cli.CommandLine;
 using Microsoft.DotNet.Cli.Utils;
-using Microsoft.DotNet.Tools.MSBuild;
-using Microsoft.DotNet.Tools;
 using Microsoft.DotNet.Cli;
-using System.Diagnostics;
 using Parser = Microsoft.DotNet.Cli.Parser;
 using System;
 using System.CommandLine.Parsing;
@@ -18,11 +15,9 @@ namespace Microsoft.DotNet.Tools.Pack
     {
         public PackCommand(
             IEnumerable<string> msbuildArgs,
-            IEnumerable<string> userDefinedArguments,
-            IEnumerable<string> trailingArguments,
             bool noRestore,
             string msbuildPath = null)
-            : base(msbuildArgs, userDefinedArguments, trailingArguments, noRestore, msbuildPath)
+            : base(msbuildArgs, noRestore, msbuildPath)
         {
         }
 
@@ -47,8 +42,6 @@ namespace Microsoft.DotNet.Tools.Pack
 
             return new PackCommand(
                 msbuildArgs,
-                parseResult.OptionValuesToBeForwarded(PackCommandParser.GetCommand()),
-                parseResult.ValueForArgument<IEnumerable<string>>(PackCommandParser.SlnOrProjectArgument) ?? Array.Empty<string>(),
                 noRestore,
                 msbuildPath);
         }

--- a/src/Cli/dotnet/commands/dotnet-publish/Program.cs
+++ b/src/Cli/dotnet/commands/dotnet-publish/Program.cs
@@ -15,11 +15,9 @@ namespace Microsoft.DotNet.Tools.Publish
     {
         private PublishCommand(
             IEnumerable<string> msbuildArgs,
-            IEnumerable<string> userDefinedArguments,
-            IEnumerable<string> trailingArguments,
             bool noRestore,
             string msbuildPath = null)
-            : base(msbuildArgs, userDefinedArguments, trailingArguments, noRestore, msbuildPath)
+            : base(msbuildArgs, noRestore, msbuildPath)
         {
         }
 
@@ -52,8 +50,6 @@ namespace Microsoft.DotNet.Tools.Publish
 
             return new PublishCommand(
                 msbuildArgs,
-                parseResult.OptionValuesToBeForwarded(PublishCommandParser.GetCommand()),
-                parseResult.ValueForArgument<IEnumerable<string>>(PublishCommandParser.SlnOrProjectArgument) ?? Array.Empty<string>(),
                 noRestore,
                 msbuildPath);
         }

--- a/src/Cli/dotnet/commands/dotnet-run/RunCommand.cs
+++ b/src/Cli/dotnet/commands/dotnet-run/RunCommand.cs
@@ -178,8 +178,6 @@ namespace Microsoft.DotNet.Tools.Run
             var buildResult =
                 new RestoringCommand(
                     restoreArgs.Prepend(Project),
-                    restoreArgs,
-                    new[] { Project },
                     NoRestore
                 ).Execute();
 

--- a/src/Cli/dotnet/commands/dotnet-test/Program.cs
+++ b/src/Cli/dotnet/commands/dotnet-test/Program.cs
@@ -15,11 +15,9 @@ namespace Microsoft.DotNet.Tools.Test
     {
         public TestCommand(
             IEnumerable<string> msbuildArgs,
-            IEnumerable<string> userDefinedArguments,
-            IEnumerable<string> trailingArguments,
             bool noRestore,
             string msbuildPath = null)
-            : base(msbuildArgs, userDefinedArguments, trailingArguments, noRestore, msbuildPath)
+            : base(msbuildArgs, noRestore, msbuildPath)
         {
         }
 
@@ -64,8 +62,6 @@ namespace Microsoft.DotNet.Tools.Test
 
             TestCommand testCommand = new TestCommand(
                 msbuildArgs,
-                result.OptionValuesToBeForwarded(TestCommandParser.GetCommand()),
-                result.ValueForArgument<IReadOnlyCollection<string>>(TestCommandParser.SlnOrProjectArgument),
                 noRestore,
                 msbuildPath);
 

--- a/src/Tests/dotnet.Tests/dotnet-msbuild/GivenDotnetBuildInvocation.cs
+++ b/src/Tests/dotnet.Tests/dotnet-msbuild/GivenDotnetBuildInvocation.cs
@@ -54,6 +54,8 @@ namespace Microsoft.DotNet.Cli.MSBuild.Tests
 
         [Theory]
         [InlineData(new string[] { "-f", "tfm" }, "-target:Restore", "-property:TargetFramework=tfm")]
+        [InlineData(new string[] { "-p:TargetFramework=tfm" }, "-target:Restore", "-property:TargetFramework=tfm")]
+        [InlineData(new string[] { "-t:Run", "-f", "tfm" }, "-target:Restore", "-property:TargetFramework=tfm -t:Run")]
         [InlineData(new string[] { "-o", "myoutput", "-f", "tfm", "-v", "diag", "/ArbitrarySwitchForMSBuild" },
                                   "-target:Restore -verbosity:diag -property:OutputPath=<cwd>myoutput /ArbitrarySwitchForMSBuild",
                                   "-property:TargetFramework=tfm -verbosity:diag -property:OutputPath=<cwd>myoutput /ArbitrarySwitchForMSBuild")]

--- a/src/Tests/dotnet.Tests/dotnet-msbuild/GivenDotnetBuildInvocation.cs
+++ b/src/Tests/dotnet.Tests/dotnet-msbuild/GivenDotnetBuildInvocation.cs
@@ -54,8 +54,10 @@ namespace Microsoft.DotNet.Cli.MSBuild.Tests
 
         [Theory]
         [InlineData(new string[] { "-f", "tfm" }, "-target:Restore", "-property:TargetFramework=tfm")]
-        [InlineData(new string[] { "-p:TargetFramework=tfm" }, "-target:Restore", "-property:TargetFramework=tfm")]
+        [InlineData(new string[] { "-p:TargetFramework=tfm" }, "-target:Restore", "-p:TargetFramework=tfm")]
+        [InlineData(new string[] { "/p:TargetFramework=tfm" }, "-target:Restore", "-property:TargetFramework=tfm")]
         [InlineData(new string[] { "-t:Run", "-f", "tfm" }, "-target:Restore", "-property:TargetFramework=tfm -t:Run")]
+        [InlineData(new string[] { "/t:Run", "-f", "tfm" }, "-target:Restore", "-property:TargetFramework=tfm /t:Run")]
         [InlineData(new string[] { "-o", "myoutput", "-f", "tfm", "-v", "diag", "/ArbitrarySwitchForMSBuild" },
                                   "-target:Restore -verbosity:diag -property:OutputPath=<cwd>myoutput /ArbitrarySwitchForMSBuild",
                                   "-property:TargetFramework=tfm -verbosity:diag -property:OutputPath=<cwd>myoutput /ArbitrarySwitchForMSBuild")]

